Source/WebKit/UIProcess/API/glib/WebKitImage.cpp

```diff
@@ -0,0 +1,477 @@
+}
+#endif
+
+static guint webkit_image_hash(GIcon*);
+static gboolean webkit_image_equal(GIcon*, GIcon*);
+
+static GInputStream* webkit_image_load(GLoadableIcon*, int, char**, GCancellable*, GError**);
+static void webkit_image_load_async(GLoadableIcon*, int, GCancellable*, GAsyncReadyCallback, gpointer);
+static GInputStream* webkit_image_load_finish(GLoadableIcon*, GAsyncResult*, char**, GError**);
+
+static void webkit_image_gicon_interface_init(GIconIface* iface)
+{
+    iface->hash = webkit_image_hash;
+    iface->equal = webkit_image_equal;
+}
+
+static void webkit_image_gloadable_icon_interface_init(GLoadableIconIface* iface)
+{
+    iface->load = webkit_image_load;
+    iface->load_async = webkit_image_load_async;
+    iface->load_finish = webkit_image_load_finish;
+}
+
+guint webkit_image_hash(GIcon* icon)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(icon), 0);
+
+    auto* image = WEBKIT_IMAGE(icon);
+
+    Hasher hasher;
+
+    auto* bytes = webkit_image_as_rgba_bytes(image);
+    gsize size = 0;
+    const guint8* data = static_cast<const guint8*>(g_bytes_get_data(bytes, &size));
+    gsize rowBytes = image->priv->width * static_cast<uint8_t>(ImageFormatBPP::RGBA8);
+
+    ASSERT(size >= image->priv->stride * (image->priv->height - 1) + rowBytes);
+
+    for (int y = 0; y < image->priv->height; ++y) {
+        auto span = std::span<const uint8_t>(data + y * image->priv->stride, rowBytes);
+        WTF::add(hasher, span);
+    }
+
+    WTF::add(hasher, image->priv->width, image->priv->height);
+    return hasher.hash();
+}
+
+gboolean webkit_image_equal(GIcon* icon1, GIcon* icon2)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(icon1), false);
+
+    if (!WEBKIT_IS_IMAGE(icon2))
+        return false;
+
+    auto* image1 = WEBKIT_IMAGE(icon1);
+    auto* image2 = WEBKIT_IMAGE(icon2);
+
+    if (image1->priv->width != image2->priv->width
+        || image1->priv->height != image2->priv->height)
+        return false;
+
+    auto* bytes1 = webkit_image_as_rgba_bytes(image1);
+    auto* bytes2 = webkit_image_as_rgba_bytes(image2);
+
+    gsize size1 = 0;
+    gsize size2 = 0;
+    const guint8* data1 = static_cast<const guint8*>(g_bytes_get_data(bytes1, &size1));
+    const guint8* data2 = static_cast<const guint8*>(g_bytes_get_data(bytes2, &size2));
+
+    gsize rowBytes = image1->priv->width * static_cast<uint8_t>(ImageFormatBPP::RGBA8);
+
+    ASSERT(size1 >= image1->priv->stride * (image1->priv->height - 1) + rowBytes);
+    ASSERT(size2 >= image2->priv->stride * (image1->priv->height - 1) + rowBytes);
+
+    for (int y = 0; y < image1->priv->height; ++y) {
+        auto span1 = std::span<const uint8_t>(data1 + y * image1->priv->stride, rowBytes);
+        auto span2 = std::span<const uint8_t>(data2 + y * image2->priv->stride, rowBytes);
+
+        if (!equalSpans(span1, span2))
+            return false;
+    }
+
```

Comment:
I think this could be written using `consumeSpan()` and `span::first()`, in a similar way as suggested for the implementation of the `.hash` vfunc.